File: android/app/src/main/java/com/openpasskey/terminal/printing/BaseScanExplorer.kt
package com.openpasskey.terminal.printing

object BaseScanExplorer {
    private val transactionHash = Regex("^0x[0-9a-fA-F]{64}$")

    fun transactionUrl(chainId: Long, hash: String): String {
        require(transactionHash.matches(hash)) { "Payment transaction hash is not canonical" }
        val origin = when (chainId) {
            8453L -> "https://basescan.org"
            84532L -> "https://sepolia.basescan.org"
            else -> throw IllegalArgumentException("BaseScan is unavailable for chain $chainId")
        }
        return "$origin/tx/${hash.lowercase()}"
    }
}

File: ios/App/Sources/BaseScanExplorer.swift
import Foundation
import OPKTerminalCore

enum BaseScanExplorerError: LocalizedError, Equatable {
    case invalidTransactionHash
    case unsupportedChain(UInt64)

    var errorDescription: String? {
        switch self {
        case .invalidTransactionHash:
            "Payment transaction hash is not canonical."
        case let .unsupportedChain(chainID):
            "BaseScan is unavailable for chain \(chainID)."
        }
    }
}

enum BaseScanExplorer {
    static func transactionURL(chainID: UInt64, hash: String) throws -> URL {
        guard (try? Bytes32(hex: hash)) != nil else {
            throw BaseScanExplorerError.invalidTransactionHash
        }
        let origin: String
        switch chainID {
        case 8_453:
            origin = "https://basescan.org"
        case 84_532:
            origin = "https://sepolia.basescan.org"
        default:
            throw BaseScanExplorerError.unsupportedChain(chainID)
        }
        guard let url = URL(string: "\(origin)/tx/\(hash.lowercased())") else {
            throw BaseScanExplorerError.invalidTransactionHash
        }
        return url
    }
}
